[Soup] Attempt a content-type lookup for known resources

spenap opened the pull request:
#### 7958c632dfcf8e36c88b2fb041f67d30a5caac62
<pre>
[Soup] Attempt a content-type lookup for known resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=295698">https://bugs.webkit.org/show_bug.cgi?id=295698</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

In some platforms, shared-mime-info might not be available, in which
case g_content_type_guess will silently fail, falling back to
&quot;text/plain&quot;, without setting result_uncertain to true.

This approach uses MIMETypeRegistry to get the MIME type based on the
path, falling back to g_content_type_guess when that fails.

* Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
(WebCore::ResourceLoader::loadGResource):

Canonical link: <a href="https://commits.webkit.org/297585@main">https://commits.webkit.org/297585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff8f505f849e6720be9ea07b722be6e7642aad67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118313 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62561 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114198 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40531 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85262 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115183 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65692 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19122 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62158 "") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95410 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121640 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39310 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94078 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97228 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93902 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24009 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Michael Catanzaro and Adrian Perez de Castro; Running compile-webkit; Running canonicalize-commit; Pushed commit to WebKit repository") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39133 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16931 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35338 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39198 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44686 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38833 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42170 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40576 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/916a33472f24109617cb0154ce5e48ca1fd67fc0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47815

aperezdc commented on Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
```diff
@@ -33,12 +33,45 @@
+    if (dotIndex != notFound && dotIndex < (fileName.length() - 1))
+        extension = StringView(fileName).substring(dotIndex);
+
+    static constexpr std::pair<ComparableLettersLiteral, ASCIILiteral> pdfjsResourceMapping[] = {
```
This is not used only for PDFjs, let's call it something else, for example `extensionTypeMapping`.

mcatanzaro replied:
We already have a better version of this in MIMETypeRegistry.cpp. Access it by calling `MIMETypeRegistry::mimeTypeForExtension`.

aperezdc commented on Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
```diff
@@ -33,12 +33,45 @@
 
 namespace WebCore {
 
+static std::optional<String> contentTypeLookUpForKnownResource(const char* filename)
```
IMO the name of the function does not match what it does, which is plainly getting the MIME type from the file extension, from a fixed set of built-in mappings. How about something like “contentTypeForKnownExtension”?

```suggestion
static std::optional<String> contentTypeForKnownExtension(const char* filename)
```

aperezdc reviewed (approved):
LGTM, and I think we can merge it as-is.

This will be handy to make sure Web Inspector and PDFjs always work regardless of how odd the rest of the system where WebKit runs might be. We have had issues in the past with the inspector not working due to missing `shared-mime-info`, or regressions in it, which would have been avoided if we had done this change before.

The issue with `GResource` is that automatically-registered resource bundles are part of a global lookup mechanism which includes the resources of _all_ linked libraries and the application, plus resources manually registered by the programmer with `g_resources_register()`&mdash;and the ones loaded later can override each other. But this is already happening without this patch, so I think improvements to how we handle resource bundles can be done separately in follow-up patches.

mcatanzaro commented on Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
```diff
@@ -63,12 +96,15 @@ void ResourceLoader::loadGResource()
-        GUniquePtr<char> contentType(g_content_type_guess(fileName.get(), data, dataSize, nullptr));
-        auto contentTypeString = String::fromLatin1(contentType.get());
-        ResourceResponse response { WTFMove(url), extractMIMETypeFromMediaType(contentTypeString), static_cast<long long>(dataSize), extractCharsetFromMediaType(contentTypeString).toString() };
+        auto contentTypeString = contentTypeLookUpForKnownResource(fileName.get());
```
Unfortunately it is common that the file extension is a lie and the actual content type is something else. It's especially common for image files, where a resource will use a commonly-supported file extension like .png but the server might actually send something more sophisticated; we've seen it be JPEG 2000 in the past, and I wouldn't be surprised if nowadays something nicer like AVIF gets used instead. Another common lie is to serve an HTTP page (like Anubis) instead of whatever file you think you should be getting.

So I would reverse the order here: call `g_content_type_guess()` first, and only fall back to the file extension map (`MIMETypeRegistry::mimeTypeForExtension`) only if that fails, rather than trusting the file extension first.

aperezdc replied:
@mcatanzaro I think you are misunderstanding here. There is no network loads involved in this code path, there is no server sending anything. This is for **local**, **bundled** resources (through `GResource`). The resources we pack inside the bundle (remote inspector, PDFjs) we *know* to be “good” in that regard, because we are the ones packaging them.

The follow-up suggested is *precisely* intended to tighten the usage of the fixed extension-to-MIME-type mapping to use it only for the specific `GResource` bundle that contains the WebKit resources, so the extension is only ever used for those.

mcatanzaro replied:
Ah, fair point. So this seems fine, then.

> Even when g_content_type_guess returns text/plain it doesn't seem to set result_uncertain to TRUE (which would be useful, I guess, to tell real plain text results from failed ones).

I think it's a bug in `g_content_type_guess()`... or maybe two bugs. "text/plain" is not actually used as a fallback for everything, but only for stuff that passes the `looks_like_text()` check. We should probably just set the `result_uncertain` flag:

```
diff --git a/gio/gcontenttype-fdo.c b/gio/gcontenttype-fdo.c
index 230cea182..9d1b81d7c 100644
--- a/gio/gcontenttype-fdo.c
+++ b/gio/gcontenttype-fdo.c
@@ -625,7 +625,11 @@ g_content_type_guess_impl (const gchar  *filename,
       if (sniffed_mimetype == XDG_MIME_TYPE_UNKNOWN &&
           data &&
           looks_like_text (data, data_size))
-        sniffed_mimetype = "text/plain";
+        {
+          sniffed_mimetype = "text/plain";
+          if (result_uncertain)
+            *result_uncertain = TRUE;
+        }
 
       /* For security reasons we don't ever want to sniff desktop files
        * where we know the filename and it doesn't have a .desktop extension.
```

Also, the `looks_like_text()` function seems _really_ sketchy and is clearly not reliable. The data is detected as binary if it contains an ASCII control code other than space or `\b`, and as text otherwise. I'm not sure this is a good heuristic.

But I guess none of that matters if you check filename first.

>  Would you suggest we always call MIMETypeRegistry::mimeTypeForExtension when we get text/plain, both here and in the RemoteInspector case?

Looks like we don't need to touch RemoteInspector at all.

spenap replied:
I've used `MIMETypeRegistry::mimeTypeForPath` instead, as it extracts the extension and then invokes `MIMETypeRegistry::mimeTypeForExtension`.

mcatanzaro reviewed (changes requested):
I see RemoteInspectorHTTPServer.cpp also uses `g_content_type_guess()`. Maybe there is opportunity for a little code sharing to ensure it does the same thing, although unfortunately I'm not sure where we would put the code.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/41ebadfd6cda4b51a655a052fc2b7aebde7d8386. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47815

mcatanzaro reviewed (approved):
I think this can land with the comment removed, since the comment does not seem useful.

Normally we use `::fromUTF8` to create strings, but in these cases `fromLatin1` should be fine because:

 * The filesystem path is in filesystem encoding. We don't have any _correct_ way to construct a `StringView` that represents filesystem encoding, but we know in practice that `::fromLatin1` is not going to perform conversions or mess anything up.
 * MIME types are guaranteed to contain only ASCII characters. I might use `::fromUTF8` for the MIME type anyway, but it surely doesn't matter.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff8f505f849e6720be9ea07b722be6e7642aad67. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47815

spenap committed:
[Soup] Attempt a content-type lookup for known resources
https://bugs.webkit.org/show_bug.cgi?id=295698

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

In some platforms, shared-mime-info might not be available, in which
case g_content_type_guess will silently fail, falling back to
"text/plain", without setting result_uncertain to true.

This approach uses MIMETypeRegistry to get the MIME type based on the
path, falling back to g_content_type_guess when that fails.

* Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
(WebCore::ResourceLoader::loadGResource):

Canonical link: https://commits.webkit.org/297585@main
